Mask Windows service password in the logs

jeddcuison opened the pull request:
The masking is done via Log4j layout

akinny commented on jwala-services/src/test/java/com/cerner/jwala/service/custom/logging/WinSvcPasswordMaskingLayoutTest.java:
```diff
@@ -0,0 +1,103 @@
+public class WinSvcPasswordMaskingLayoutTest {
+
+    private static final String LOG_MSG_WITH_PASSWORD_1 = "Executing command \"~/.jwala/CTO-N9SF-LTST-HEALTH-CHECK-4.0-" +
+            "USMLVV1CDS0049-2/install-service.sh CTO-N9SF-LTST-HEALTH-CHECK-4.0-USMLVV1CDS0049-2 D:/ctp/app/instances " +
```
Can you take out all our server names

akinny commented on jwala-services/src/test/java/com/cerner/jwala/service/custom/logging/WinSvcPasswordMaskingLayoutTest.java:
```diff
@@ -0,0 +1,103 @@
+    private static final String LOG_MSG_WITH_PASSWORD_1 = "Executing command \"~/.jwala/CTO-N9SF-LTST-HEALTH-CHECK-4.0-" +
+            "USMLVV1CDS0049-2/install-service.sh CTO-N9SF-LTST-HEALTH-CHECK-4.0-USMLVV1CDS0049-2 D:/ctp/app/instances " +
+            "apache-tomcat-7.0.55 \"jedi\" the-password \"";
+    private static final String LOG_MSG_WITH_PASSWORD_2 = "D:\\cygwin64\\home\\N9SFGLabTomcatAdmin>set svc_password=the-password ";
```
remove internal user name

akinny commented on jwala-services/src/main/java/com/cerner/jwala/service/custom/logging/WinSvcPasswordMaskingLayout.java:
```diff
@@ -0,0 +1,84 @@
+        if (StringUtils.indexOf(msg, "install-service.sh") > 0) {
+            int index = msg.lastIndexOf(' ', msg.lastIndexOf(' ') - 1);
+            maskedMsg = msg.substring(0, index) + " ******** \"";
+        } else if (StringUtils.contains(msg, "set svc_password=")) {
```
Can you move the hard coded strings as constants, so it will be easy to change if required?

jeddcuison replied:
done

akinny reviewed (commented):
Do we have to change jwala/jwala-tomcat/src/main/resources/data/conf/log4j.xml for applying this override?

jeddcuison committed:
Mask svc_password in the logs

jeddcuison commented:
@akinny, I updated the log4j xmls as well as per your comment

jeddcuison commented:
Please note that I forced my latest commit to overwrite codes that have sensitive info

jeddcuison committed:
Replaced hard coded strings with constants

jmahmoodcc commented on jwala-services/src/main/java/com/cerner/jwala/service/custom/logging/WinSvcPasswordMaskingLayout.java:
```diff
@@ -0,0 +1,89 @@
+public class WinSvcPasswordMaskingLayout extends PatternLayout {
+
+    private static final String INSTALL_SERVICE_SH = "install-service.sh";
+    private static final String SET_SVC_PASSWORD = "set svc_password=";
```
Is it possible to create the above 2 constants as parameters in the log4j.xml?

I'm just thinking that these searches are very specific to the current scripts, and in case it changes in the future it would be easier to just update the log4j.xml then have to rebuild the whole project.

jeddcuison replied:
Working on it...

jmahmoodcc reviewed (commented):
Question about parameterizing the search expressions

PeterTeunissen commented:
Instead of searching for specific keywords and scrub those, I think the scrubber should have a static list of  strings that should be scrubbed. Then, when logging a line, replace all the occurrences of the strings in the list with '****'.  You populate the list of static strings whenever you call the password decrypt routine (so with the clear text pw). This way nobody can forget to change the scrubber when you change the shell scripts or other keywords....

public class Scrubber {

    private static HashMap<String, String> toRemove = new HashMap<String, String>();

    public static void addKey(String key) {
        String mask = "";
        for (int i = 0; i < key.length(); i++) {
            mask += "*";
        }
        toRemove.put(key, mask);
    }

    public static String scrubMe(String raw) {
        String clean = raw;
        for (String key : toRemove.keySet()) {
            clean = clean.replaceAll(key, toRemove.get(key));
        }
        return clean;
    }

    public static void main(String[] args) {
        System.out.println(scrubMe("My Password is: secret"));
        Scrubber.addKey("secret");
        System.out.println(scrubMe("My Password is: secret"));
    }
}

Output:
My Password is: secret
My Password is: ******

PeterTeunissen reviewed (commented):
See my comment... just my $0.02

akinny commented:
I like it Peter!! Just make the put in a synchronized block

jeddcuison commented:
Thanks @PeterTeunissen, we implemented your solution

jeddcuison committed:
Implement password scrubber by storing encrypted passwords in a set then comparing them to the log message

Other changes:

- Remove the "other" or the old password scrubber

akinny commented on jwala-services/src/main/java/com/cerner/jwala/service/custom/logging/impl/JvmWinSvcAcctPasswordScrubber.java:
```diff
@@ -0,0 +1,95 @@
+        this.decryptor = decryptor;
+
+        // Populate the map of items to remove in the logs
+        final List<Jvm> jvms = jvmPersistenceService.getJvms();
```
When is this initialized? What happens to newly added JVM's with service accounts

jmahmoodcc replied:
What if the map was populated whenever a JVM is added, updated, or removed? You'd probably have to synchronize access to it as well.

jeddcuison committed:
Update JVM Windows service password store on JVM creation, update and deletion

jeddcuison committed:
Synchronize password store

jeddcuison committed:
Replace HashSet with CopyOnWriteArraySet

CopyOnWriteArraySet makes modifications on the Set not affect iteration

jeddcuison committed:
Refactor

- Rename variable(s)
- Make applicable method array static
- Make decryptor final

jeddcuison committed:
Add TODO

jeddcuison committed:
Fix masking not being applied to the log

jeddcuison commented:
This was closed in lieu of https://github.com/cerner/jwala/pull/137
